pyprql/magic/README.md

```diff
@@ -0,0 +1,184 @@
+# Using the IPython and Jupyter Magic
+
+Work with pandas and PRQL in an IPython terminal or Jupyter notebook.
+
+## Implementation
+
+This is a this wrapper around the fantastic
+[IPython-sql][ipysql] magic.
+Roughly speaking,
+all we do is parse PRQL to SQL and pass that through to `ipython-sql`.
+A full documentation of the supported features is available at their
+[repository][ipysql].
+Here,
+we document those places where we differ from them,
+plus those features we think you are mostly likely to find useful.
+
+## Usage
+
+### Installation
+
+If you have already installed PyPRQL into your environment,
+then you should be could to go!
+We bundle in `IPython` and `pandas`,
+though you'll need to install `Jupyter` separately.
+If you haven't installed PyPRQL,
+that's as simple as:
+
+```shell
+pip install PyPRQL
+```
+
+### Set Up
+
+Open up either an `IPython` terminal or `Jupyter` notebook.
+First,
+we need to load the extension and connect to a database.
+
+```python
+In [1]: %load_ext pyprql.magic
+
+In [2]: %prql duckdb:///:memory:
+
+```
+
+When connecting to a database,
+pass the connection string as an argument to the line magic `%prql`.
+The connection string needs to be in [SQLAlchemy format][conn_str],
+so any connection supported by `SQLAlchemy` is supported by the magic.
+Additional connection parameters can be passed as a dictionary using the `--connection_arguments`
+flag to the the `%prql` line magic.
+We ship with the necessary extensions to use [DuckDB][duckdb]
+as the backend,
+and here connect to an in-memory database.
+
+However,
+in-memory databases start off empty!
+So,
+we need to add some data.
+This is where [pandas][pandas]
+comes into play.
+We can easily add a dataframe to the `DuckDB` database like so:
+
+```python
+In [3]: %prql --persist data
+
+```
+
+where data is a pandas dataframe we have already loaded.
+This adds a table named `data` to the in-memory `DuckDB` instance.
+If you connect to an existing SQL database,
+then all the tables normally there will be accessible.
+
+### Usage
+
+:::{Important}
+This is one area where we differe from `IPython-sql`.
+We only support parasing PRQL as a cell magic,
+not as a line magic.
```

Comment:
I'm a bit confused by this as in line 41 above you seem to have a line magic example, or is that passing a parameter to the `prql` magic which is then used in the cell magics further on?

Reply:
So we do have a line magic, but it can only be used for connection strings and parameters. In the original Python-sql, line magic sql is valid, like this `%sql select * from table`. But they have quite an extensive parser to figure out whether the line magic is a connection string, flags, or sql. To the first order of approximation, they achieve this by checking to see if the line passed looks like SQL. So, if we wanted this functionality, we'd have to re-implement their parser, look for PRQL, parse the PRQL to SQL, re-assemble the line, then pass it to the magic. This seemed a bit of a faf, so for the momentI've limited the line magic to parameters and connections strings. I'll clarify in the docs!